### PR TITLE
PYTHON-2884: Replaced SON usage in all internal classes and commands

### PR DESCRIPTION
**Pull request by Jibola**
## Callouts 
* Pool.metadata is a _somewhat_ public method that is now returning `SON` instead of `dict`. I would love perspective on if this could be considered a breaking conversion.
* There are two test cases that now needed changing. Commentary needed on whether or not these are breaking changes.

### CODE REVIEW

**Comment by blink1073** [comment]
> I would love perspective on if this could be considered a breaking conversion.

Yeah I think to be safe we should leave it as SON until 5.0.  It looks like some tests need to be updated as well.

**Comment by ShaneHarvey** [comment]
Pool is a private API so there's no compat concerns there,

**Comment by blink1073** [comment]
> Pool is a private API so there are no compat concerns.

Ah, true, I missed that `MongoClient.__options` is private, I saw `options.pool_options.metadata` in a grep.

**Comment by ShaneHarvey** [comment]
Oh I didn't realize it was PoolOptions. That is indeed public:
```python
>>> client.options.pool_options.metadata
SON([('driver', SON([('name', 'PyMongo'), ('version', '4.7.0.dev0')])), ('os', SON([('type', 'Darwin'), ('name', 'Darwin'), ('architecture', 'arm64'), ('version', '13.6')])), ('platform', 'CPython 3.11.4.final.0')])
```

**Comment by blink1073** [comment]
Looks like two more tests need to be addressed:

```
FAILED test/test_collection.py::TestCollection::test_find_command_generation
FAILED test/test_cursor.py::TestCursor::test_clone - AssertionError: False is...
```

**Review by blink1073 — APPROVED** [review]
LGTM!

**Comment by ShaneHarvey** [comment]
Please wait for my review before merging.

File: pymongo/message.py
Comment (ShaneHarvey) [inline pymongo/message.py:397]: Let's change _encrypter.encrypt's return type to dict so we can remove the redundant casts here.

File: pymongo/pool.py
Comment (ShaneHarvey) [inline pymongo/pool.py:289]: Odd that this shows up in this PR. Did you `git merge master`?

File: test/test_collection.py
Comment (ShaneHarvey) [inline test/test_collection.py:2128]: Hmm, this highlights how this is a subtle breaking change since the command payloads published in CommandStartedEvent will change from SON to dict. Code relying on SON will break when it changes to dict.

We should at least document in the changelog and mark the jira ticket as a minor breaking change.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: test/test_collection.py
Comment (ShaneHarvey) [reply test/test_collection.py:2128]: Also since this is the case I revise my earlier opinion. I think it's fine to change PoolOptions.metadata type.

File: pymongo/encryption.py
Comment (NoahStapp) [inline pymongo/encryption.py:26]: Should this be lowercase `dict` for consistency across type hints?

File: test/test_collection.py
Comment (NoahStapp) [reply test/test_collection.py:2128]: If we're removing SON entirely, this test case should compare against an actual `dict`, not a converted `SON`.

**Review by NoahStapp — CHANGES_REQUESTED** [review]
A few minor comments, otherwise looks great, thanks Jib! Also, if we're removing `SON` entirely, can we also delete the `bson.son` file?

**Comment by Jibola** [comment]
> A few minor comments, otherwise looks great, thanks Jib! Also, if we're removing `SON` entirely, can we also delete the `bson.son` file?

Ah, I don't think we're removing`SON` entirely, we're just deprecating our internal use of it. IIUC, SON is still perfectly viable, but outright removal would be a major version breaking change.

File: pymongo/encryption.py
Comment (Jibola) [reply pymongo/encryption.py:26]: This was the error I ran into, which left me just to leave it. 
```
typecheck: commands[1]> mypy --install-types --non-interactive bson gridfs tools pymongo
pymongo/encryption.py:86: error: "dict" is not subscriptable  [misc]
        document_class=dict[str, Any], uuid_representation=STANDARD
                       ^
Found 1 error in 1 file (checked 86 source files)
```

Comment (NoahStapp) [reply pymongo/encryption.py:26]: We're importing future `__annotations__`, so I would expect that to work, hmm.

**Review by NoahStapp — APPROVED** [review]
Looks good! I'd prefer to figure out the `Dict` vs `dict` issue in `encryption.py`, but otherwise LGTM!

File: pymongo/encryption.py
Comment (blink1073) [inline pymongo/encryption.py:86]: This `Dict` is a runtime object, that's what the annotations import doesn't help.

Comment (NoahStapp) [reply pymongo/encryption.py:86]: Ahhh right, we're passing it directly as a keyword argument.

File: doc/changelog.rst
Comment (ShaneHarvey) [inline doc/changelog.rst:14]: Let's call this out more prominently under an "Unavoidable breaking changes" section like we do here: https://github.com/mongodb/mongo-python-driver/blob/3d63eedb/doc/changelog.rst?plain=1#L291

It's also best practice to show how this breaking change could show up and what to do about it. Including example code that expects `CommandStartedEvent.command` to be SON would be good.

Comment (ShaneHarvey) [reply doc/changelog.rst:14]: https://github.com/mongodb/mongo-python-driver/pull/1498

**Pull request merged**
